src/libteam/0005-libteam-Add-warm_reboot-mode.patch

```diff
@@ -0,0 +1,354 @@
+ 	if (!lacp_port->periodic_on &&
+-	    memcmp(&lacpdu.partner, &lacp_port->actor,
++	    memcmp(&lacpdu->partner, &lacp_port->actor,
+ 		   sizeof(struct lacpdu_info))) {
+ 		err = lacpdu_send(lacp_port);
+ 		if (err)
+@@ -1133,6 +1133,47 @@ static int lacpdu_recv(struct lacp_port *lacp_port)
+ 	return 0;
+ }
+ 
++static int lacpdu_recv(struct lacp_port *lacp_port)
++{
++	struct lacpdu lacpdu;
++	struct sockaddr_ll ll_from;
++	int err;
++
++	err = teamd_recvfrom(lacp_port->sock, &lacpdu, sizeof(lacpdu), 0,
++			     (struct sockaddr *) &ll_from, sizeof(ll_from));
++	if (err <= 0)
++		return err;
++
++	return lacpdu_process(lacp_port, &lacpdu);
++}
++
++static int lacpdu_read(struct lacp_port *lacp_port)
++{
++	FILE* fp;
++	char path[PATH_MAX];
++	struct lacpdu lacpdu;
++	int err;
++
++	strcpy(path, lacp_port->ctx->lacp_directory);
++	if (path[strlen(path) - 1] != '/')
++		strcat(path, "/"); /* Add trailing slash if we don't have one in the path */
++	strcat(path, lacp_port->tdport->ifname);
++	fp = fopen(path, "r");
++	if (!fp) {
++		teamd_log_err("Can't open lacp-saved dump from file %s: %s", path, strerror(errno));
++		return errno;
++	}
++
++	err = fread(&lacpdu, sizeof(struct lacpdu), 1, fp);
++	(void)fclose(fp);
++	if (err <= 0) {
++		teamd_log_err("Can't read lacp-saved dump from file %s: %s", path, strerror(errno));
++		return err;
++	}
++
++        return lacpdu_process(lacp_port, &lacpdu);
++}
++
+ static int lacp_callback_timeout(struct teamd_context *ctx, int events,
+ 				 void *priv)
+ {
+@@ -1299,6 +1340,11 @@ static int lacp_port_added(struct teamd_context *ctx,
+ 	lacp_port_actor_init(lacp_port);
+ 	lacp_port_link_update(lacp_port);
+ 
++	/* Read data from file and process it */
++	if (ctx->warm_reboot && ctx->lacp_directory) {
++		(void)lacpdu_read(lacp_port);
++	}
++
+ 	teamd_loop_callback_enable(ctx, LACP_SOCKET_CB_NAME, lacp_port);
+ 	return 0;
+ 
+@@ -1321,7 +1367,11 @@ static void lacp_port_removed(struct teamd_context *ctx,
+ {
+ 	struct lacp_port *lacp_port = priv;
+ 
+-	lacp_port_set_state(lacp_port, PORT_STATE_DISABLED);
++	if (!lacp_port->ctx->keep_ports) {
++		/* Don't transition into DISABLED state,
++		   which sends EXPIRED LACP PDU update */
++		lacp_port_set_state(lacp_port, PORT_STATE_DISABLED);
++	}
+ 	teamd_loop_callback_del(ctx, LACP_TIMEOUT_CB_NAME, lacp_port);
+ 	teamd_loop_callback_del(ctx, LACP_PERIODIC_CB_NAME, lacp_port);
+ 	teamd_loop_callback_del(ctx, LACP_SOCKET_CB_NAME, lacp_port);
+@@ -1413,6 +1463,29 @@ static void lacp_event_watch_refresh(struct teamd_context *ctx, struct teamd_por
+ 		(void) lacpdu_send(lacp_port);
+ }
+ 
++static void lacp_event_watch_port_flush_data(struct teamd_context *ctx, struct teamd_port *tdport, void *priv)
++{
++	struct lacp *lacp = priv;
++
++	struct lacp_port *lacp_port = lacp_port_get(lacp, tdport);
++	if(lacp_port->lacpdu_saved && lacp_port->ctx->lacp_directory) {
++		char filename[PATH_MAX];
++		strcpy(filename, lacp_port->ctx->lacp_directory);
```

Comment:
in case the directory is not ended with '/', you may want to add here.

Reply:
Addressed